rust/lance-encoding/src/encodings/physical/bitpack.rs

```diff
@@ -0,0 +1,662 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use std::sync::Arc;
+
+use arrow::array::ArrayData;
+use arrow::datatypes::{ArrowPrimitiveType, UInt16Type, UInt32Type, UInt64Type, UInt8Type};
+use arrow_array::{cast::AsArray, Array, ArrayRef, PrimitiveArray};
+use arrow_schema::DataType;
+use bytes::{Bytes, BytesMut};
+use futures::future::{BoxFuture, FutureExt};
+use log::trace;
+use num_traits::{AsPrimitive, PrimInt};
+use snafu::{location, Location};
+
+use lance_arrow::DataTypeExt;
+use lance_core::{Error, Result};
+
+use crate::encoder::EncodedBufferMeta;
+use crate::{
+    decoder::{PageScheduler, PrimitivePageDecoder},
+    encoder::{BufferEncoder, EncodedBuffer},
+};
+
+// Compute the number of bits to use for each item, if this array can be encoded using
+// bitpacking encoding. Returns `None` if the type or array data is not supported.
+pub fn num_compressed_bits(arr: ArrayRef) -> Option<u64> {
+    match arr.data_type() {
+        DataType::UInt8 => num_bits_for_type::<UInt8Type>(arr.as_primitive()),
+        DataType::UInt16 => num_bits_for_type::<UInt16Type>(arr.as_primitive()),
+        DataType::UInt32 => num_bits_for_type::<UInt32Type>(arr.as_primitive()),
+        DataType::UInt64 => num_bits_for_type::<UInt64Type>(arr.as_primitive()),
+        // TODO -- eventually we could support signed types as well
+        _ => None,
+    }
+}
+
+// Compute the number of bits to to use for bitpacking generically.
+// Returns `None` if the array is empty or only contains null values.
+fn num_bits_for_type<T>(arr: &PrimitiveArray<T>) -> Option<u64>
+where
+    T: ArrowPrimitiveType,
+    T::Native: PrimInt + AsPrimitive<u64>,
+{
+    let max = arrow::compute::bit_or(arr);
+    let num_bits: Option<u64> =
+        max.map(|x| arr.data_type().byte_width() as u64 * 8 - x.leading_zeros() as u64);
+
+    // we can't bitpack into 0 bits, so the minimum is 1
+    num_bits.map(|x| x.max(1))
+}
+
+#[derive(Debug)]
+pub struct BitpackingBufferEncoder {
+    num_bits: u64,
+}
+
+impl BitpackingBufferEncoder {
+    pub fn new(num_bits: u64) -> Self {
+        Self { num_bits }
+    }
+}
+
+impl BufferEncoder for BitpackingBufferEncoder {
+    fn encode(&self, arrays: &[ArrayRef]) -> Result<(EncodedBuffer, EncodedBufferMeta)> {
+        // calculate the total number of bytes we need to allocate for the destination.
+        // this will be the number of items in the source array times the number of bits.
+        let count_items = count_items_to_pack(arrays);
+        // TODO: make function for count & round up
+        let mut dst_bytes_total = count_items * self.num_bits as usize / 8;
+        // if if there's a partial byte at the end, we need to allocate one more byte
+        if (count_items * self.num_bits as usize) % 8 != 0 {
+            dst_bytes_total += 1;
+        }
```

Comment:
FWIW, I think there is `bit_util::ceil_div` in one of the low-level arrow crates (maybe `arrow-buffer`?)

Reply:
Sounds good, made this change.